File: java/L1000/LeetCode1014.java
package L1000;

public class LeetCode1014 {
    /**
     * 由 ans = A[i] + A[j] + i - j
     * 得到：ans = (A[i] + i) + (A[j] - j)
     */
    public int maxScoreSightseeingPair(int[] A) {
        int ans = 0;
        int aii = A[0] + 0;
        for (int i = 1;i < A.length;++i) {
            ans = Math.max(ans, aii + A[i] - i);
            if (A[i] + i > aii) {
                aii = A[i] + i;
            }            
        }
        return ans;
    }
}

File: cpp/L1000/LeetCode1014.cc
#include <vector>
using std::vector;

class Solution {
public:
    int maxScoreSightseeingPair(vector<int>& values) {
        int flow{-1}, fhigh{-1};
        int slow{-1}, shigh{-1};
        int N = values.size();
        int beforeMax = -0x3f3f3f;
        int ans = -0x3f3f3f;
        for (int i = 0;i < N;++i) {
            ans = std::max(ans, beforeMax + values[i] - i);
            beforeMax = std::max(beforeMax, values[i] + i);
        }
        return ans;
    }
};

int main() {
    Solution s;
    vector<int> v{2,2,2};
    s.maxScoreSightseeingPair(v);
}